src/Microsoft.DotNet.Interactive.HttpRequest.Tests/ParserTests.cs

```diff
@@ -0,0 +1,645 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Linq;
+using FluentAssertions;
+using FluentAssertions.Execution;
+using Microsoft.DotNet.Interactive.HttpRequest.Tests.Utility;
+using Xunit;
+
+namespace Microsoft.DotNet.Interactive.HttpRequest.Tests;
+
+public class ParserTests
```

Comment:
(Possibly something to tackle in a follow up PR)

I just realized that we don't have any test that validates roundtripping for syntax trees produced by the parser. Roundtripping is an invariant property of the trees - in other words, every tree generated by the parser should preserve all the text present in the original input string and calling `tree.RootNode.TextWithTrivia` should always produce a string that is identical to the original input string.

It may be useful to validate roundtripping for some of the interesting cases being tested below (especially the ones related to leading and trailing trivia, optional syntax elements etc.) as that could alert us to future bugs in the implementation. However, adding asserts for this in the corresponding tests would violate the one-assertion-per-test guideline.

Perhaps we should add some dedicated roundtripping tests? For example, one such test could parse a string that includes all types of tokens, trivia and nodes in the grammar.

Thoughts?

Reply:
Sounds like a very good test. It might be a good candidate for a reusable FluentAssertions extension method.

Reply:
Resolving since we have now added this (in the new Parser() method).